core/src/kura.rs

```diff
@@ -201,8 +210,16 @@ impl Kura {
                                 .push(block_hash);
                         }
                         Err(error) => {
-                            error!(%error, "Failed to write block. Kura will now init again.");
-                            panic!("It is unclear what we should do here.");
+                            failed_to_store_block = Some(serialized_block);
+                            failed_to_store_block_reason = Some(error);
+                            std::thread::sleep(std::time::Duration::from_millis(250));
+                            // Why is this sleep needed?
+                            // Well in many of our tests we have kura operate within temp directories.
+                            // When Iroha is shutdown very fast these temp directories get free'd before
+                            // this thread can exit, causing a panic due to faulty I/O. This pushing of
+                            // the pushing to the next round makes sure that we don't panic if we were going
+                            // to exit anyway. The sleep is also required to make sure that the exit code
+                            // has finished running before the next iteration of the kura thread loop.
```

Comment:
This seems to be a problem with how we handle shutdown, and to me it seems that a fixed timeout isn't going to help much. 

Firstly, our tests no longer continue running on panic. Have a look at #2630. This is probably what we should do here more-less. 

Secondly, instead of expecting the exit code to finish running in a fixed amount of time, there should be a signal that the exit code finished running, not an expectation that by this time it would have.

Reply:
This is a problem with RAII and our tests. The temporary filesystem is destroyed at the same time as we shutdown iroha. In production this is not an issue. If we want to solve this properly there should be a iroha.shutdown() method that tells all the threads to exit and waits their termination. Then each test can call this at the end. Shall I create an issue for it in the "Polishing Kura" epic?